### PR TITLE
Container tagging for AWS Fargate 1.4+

### PR DESCRIPTION
**Pull request by SammyK**
### Description

This PR adds Fargate 1.4+ support to container tagging. The ID parsing patterns have been updated for consistency with other tracers (namely DataDog/dd-trace-java#2432 & DataDog/dd-trace-py#2132) and the ID parsing is now done with the POSIX regex API.

### Readiness checklist
- [x] (only for Members) Changelog has been added to the release document.
- [x] Tests added for this feature/bug.

### Reviewer checklist
- [x] Appropriate labels assigned.
- [x] Milestone is set.
- [x] Changelog has been added to the release document. For community contributors the reviewer is in charge of this task.

### CODE REVIEW

File: components/container_id/container_id.c
Comment (morrisonlevi) [inline components/container_id/container_id.c:80]: I'm not going to block this PR, but whenever there is more than 1 thing to cleanup on an early return I try to re-examine the structure to try to make it cleaner (if possible without goto). Can you put some time on my calendar to go over this?

File: components/container_id/container_id.c
Comment (morrisonlevi) [inline components/container_id/container_id.c:104]: I would expect a `memcpy` here, like the case above.

File: components/container_id/container_id.c
Comment (morrisonlevi) [inline components/container_id/container_id.c:48]: If this "fails", how does the caller know?

**Review by morrisonlevi — DISMISSED** [review]
I have a few questions. One of them might be a blocker depending on its answer but I've approved this in case it's not so we can work async; please merge responsibly :)

File: components/container_id/container_id.c
Comment (SammyK) [reply components/container_id/container_id.c:80]: The funny thing is, this is the cleaned up version; you should have seen it before. 😆 But absolutely, if you have some ideas on how to make this even cleaner, I'd love to hear your thoughts.

File: components/container_id/container_id.c
Comment (SammyK) [reply components/container_id/container_id.c:104]: There isn't an error here, but there might be a cleaner way to write this. The reason there is no `memcpy` here is because the happy path of finding a regular container ID copies directly into the main buffer of the caller (`buf`) but the edge case of a task ID uses an intermediary buffer (`task_buf`) since we'll likely have a valid container ID before finding a task ID.

But I think we can avoid the intermediary buffer by returning a "found" status from `dd_find_id_regex`. I'll rework this a bit and see if you think it's a bit cleaner.

File: components/container_id/container_id.c
Comment (SammyK) [reply components/container_id/container_id.c:48]: Given a valid `buf` is provided by the caller, it will always be filled with a container ID or an empty string. But the caller won't know if the empty string means there was an error leading up to parsing the cgroup file or if there was no valid container ID found in the file.

Do you think returning an error code would be better so that the caller could choose to emit a debug message if there was an error parsing the cgroup file?

File: components/container_id/container_id.c
Comment (morrisonlevi) [inline components/container_id/container_id.c:88]: It might be sixes, but this sort of style ends up being cleaner in some cases:

```c
char *current = start;
char *end = start + 32;
for (; current != end; ++current) {
  if (!isxdigit(*current)) break;
}
if (current != end) {
  continue;
}
```

I don't like the `LEN_SO_FAR` macro that uses local variables "magically", which is why I suggest trying it the other style.

**Review by morrisonlevi — CHANGES_REQUESTED** [review]
I'll go over it today, but this `LEN_SO_FAR` macro that uses local variables immediately jumped out at me. Suggestion below.

**Review by morrisonlevi — COMMENTED** [review]
The rest of this looks good! I'm feeling good after my nap today so I'm going to take a stab at the suggestion I made earlier to get rid of `LEN_SO_FAR`.

Also, I just want to once again say how nice it is to have self-contained components. This may not be a perfect scheme, but it's so much better than what we were doing before!

**Comment by labbati** [comment]
@SammyK the same tests fails on `master` as well so please do not consider this PR blocked once @morrisonlevi approves the latest changes. Thanks for working on this.

File: components/container_id/container_id.c
Comment (morrisonlevi) [reply components/container_id/container_id.c]: Resolved by #1194.

**Review by morrisonlevi — APPROVED** [review]

**Pull request merged**
